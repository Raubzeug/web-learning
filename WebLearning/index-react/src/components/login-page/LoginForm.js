import React from 'react';
import Message from '../Message'

class LoginForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {username: '', password: ''}
        this.handleChange = this.handleChange.bind(this)
        this.submitForm = this.submitForm.bind(this)
    }

    validateForm = () => (
        this.state.username.length > 3 && this.state.password.length > 3
    )

    // handleChange = event => {
    //     this.setState({
    //         [event.target.id]: event.target.value
    //     })
    // }

    handleChange = ({target: {name, value}}) => this.setState({ [name]: value })

    submitForm = event => {
        event.preventDefault()
        this.props.submit({
            username: this.state.username,
            password: this.state.password
        })
    }

    render = () => (
        <form className='login-form' method="POST" name='login' 
        autoComplete="on" onSubmit={this.submitForm}>
            <label className='login-form__label'>
                Логин:
                <input className='login-form__input' type='text' name='username' 
                required autoFocus 
                value={this.state.username} onChange={this.handleChange}/>
            </label>
            <label className='login-form__label'>
                Пароль:
                <input className='login-form__input' type='password' 
                name='password' required
                value={this.state.password} onChange={this.handleChange} />
            </label>

            <div className="login-form__buttons">
                <input className='login-form__submit' type="submit" value="Войти"
                disabled={!this.validateForm()} />
            </div>
            <Message success={this.props.success} errors={this.props.errors}/>
        </form>)
}

export default LoginForm