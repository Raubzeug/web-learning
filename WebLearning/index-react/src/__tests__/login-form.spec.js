import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import LoginForm from '../components/login-page/LoginForm'

describe('Login form tests', () => {
    const mockLogin = jest.fn()
        const props = {
            submit: mockLogin
        } 
    const wrapper = shallow(
        <LoginForm {...props}/>
    )
    const initialState = {
        username: '',
        password: ''
    }
    
    beforeEach(() => {
        wrapper.setState(initialState)
    })
    it('default renders properly', () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })
    it('handle state username', () => {
        wrapper.find('input[type="text"]').simulate('change', {
            target: {name: 'username', value: 'testuser'}})
        expect(wrapper.state().username).toEqual('testuser')
    })
    it('handle state password', () => {
        wrapper.find('input[type="password"]').simulate('change', {
            target: {name: 'password', value: 'testpass'}})
        expect(wrapper.state().password).toEqual('testpass')
    })
    it('form validation', () => {     
        wrapper.find('input[type="text"]').simulate('change', {
            target: {name: 'username', value: 'tes'}})
        wrapper.find('input[type="password"]').simulate('change', {
            target: {name: 'password', value: 'tes'}})
        expect(wrapper.find('.login-form__submit').first().prop('disabled')).toEqual(true)
        wrapper.find('input[type="text"]').simulate('change', {
            target: {name: 'username', value: 'tes'}})
        wrapper.find('input[type="password"]').simulate('change', {
            target: {name: 'password', value: 'testpas'}})
        expect(wrapper.find('.login-form__submit').first().prop('disabled')).toEqual(true)
        wrapper.find('input[type="text"]').simulate('change', {
            target: {name: 'username', value: 'testuser'}})
        wrapper.find('input[type="password"]').simulate('change', {
            target: {name: 'password', value: 'tes'}})
        expect(wrapper.find('.login-form__submit').first().prop('disabled')).toEqual(true)
        wrapper.find('input[type="text"]').simulate('change', {
            target: {name: 'username', value: 'testuser'},})
        wrapper.find('input[type="password"]').simulate('change', {
            target: {name: 'password', value: 'testpas'}})
        expect(wrapper.find('.login-form__submit').first().prop('disabled')).toEqual(false)
    })
    it('submitting form', () => {
        wrapper.find('input[type="text"]').simulate('change', {
            target: {name: 'username', value: 'testuser'},})
        wrapper.find('input[type="password"]').simulate('change', {
            target: {name: 'password', value: 'testpas'}})
        wrapper.find('form').simulate('submit', {
            preventDefault: () => {}
        })
        expect(mockLogin).toHaveBeenCalledTimes(1)
    })
    it('when clicking login button', () => {
        wrapper.find('input[type="text"]').simulate('change', {
            target: {name: 'username', value: 'testuser'},})
        wrapper.find('input[type="password"]').simulate('change', {
            target: {name: 'password', value: 'testpas'}})
        wrapper.find('.login-form__submit').simulate('click', {
            preventDefault: () => {}
        })
        expect(mockLogin).toHaveBeenCalledTimes(2)
    })
})