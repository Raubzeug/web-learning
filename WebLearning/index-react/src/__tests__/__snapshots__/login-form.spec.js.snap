// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form tests default renders properly 1`] = `
<form
  autoComplete="on"
  className="login-form"
  method="POST"
  name="login"
  onSubmit={[Function]}
>
  <label
    className="login-form__label"
  >
    Логин:
    <input
      autoFocus={true}
      className="login-form__input"
      name="username"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </label>
  <label
    className="login-form__label"
  >
    Пароль:
    <input
      className="login-form__input"
      name="password"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
  </label>
  <div
    className="login-form__buttons"
  >
    <input
      className="login-form__submit"
      disabled={true}
      onClick={[Function]}
      type="submit"
      value="Войти"
    />
  </div>
  <Message />
</form>
`;
